resources/templates/md/posts/2016-04-04-state.md

```diff
@@ -0,0 +1,366 @@
+```
+
+To get the current value of the agent we will dereference it the same way as an `atom`. Using `clojure.core/deref` or `@` reader macro.
+
+```clojure
+@marshmallows-in-the-bag
+
+;;=> 30
+```
+
+To update the agent value (refers agent to the new state), we will use `send` or `send-off`.
+
+They are mostly similar, the difference is in how they're implemented.  
+
+If you send the actions to any agent using `send`, they run in a thread pool with a couple of more threads than the physical number of processors. 
+If you make 1000 calls using `send`, the call that can't be processed immediately just wait until the processor becomes available.
+If they block then the thread pool can run dry.
+
+If you send the actions to any agent using `send-off`, a new thread is created for each call. 
+if you make 1000 calls using `send-off`, the call that can't be processed immediately still wait for the next available processor, 
+but they may incur the extra overhead of starting a thread if the send-off threadpool happens to be running low. 
+It's ok if the threads block because each task (potentially) gets a dedicated thread.
+
+```clojure
+(defn number-of-marshmallows-left [marshmallows-eaten]
+  (send marshmallows-in-the-bag #(- % marshmallows-eaten)))
+```
+
+So if you call the `number-of-marshmallows-left` there's a possibility that the value is still not updated.
+
+```Clojure
+@marshmallows-in-the-bag
+
+;;=> 30
+```
+Few seconds later
+
+```Clojure
+@marshmallows-in-the-bag
+
+;;=> 20
+```
+
+
+## Refs
+
+Refs are ideal for managing states of more than one identity. 
+
+For example on the llamas party there is one bag with gummy bears and one bag with marshmallows. 
+Greg and Gregorina are eating from the both bags. 
+So when the llama takes one candy, we need to update state of the llamas and state of the bag.
+
+How we can manage this states? We will use `refs`.
+
+First we need to define refs for `marshmallow`, `gummy-bears`, `greg` and `gregorina`.
+
+```Clojure
+(def marshmallow (ref {:name "marshmallow"
+                       :count-in-the-bag 30}))
+                                
+(def gummy-bears (ref {:name "gummy bears"
+                       :count-in-the-bag 20}))
+                                                                         
+(def greg (ref {:candies-eaten 0}))
+                                                                                  
+(def gregorina (ref {:candies-eaten 0}))
+```
+
+We can dereference refs the same way as we dereference atoms.
+
+```Clojue
+@greg
+
+;;=> {:candies-eaten 0}
+```
+Now we are prepared for the transfer (move one candy from the bag to llama's stomach). 
+We need to modify `candy` ref to take one candy from the bag, and modify the `llama` ref to eat one candy.
+We modify `refs` using `alter`, and we need to put all transaction operations into `dosync` 
+to initiate the transaction.
+
+```Clojure
+(defn eat-candy [name candy]
+  (if (> (:count-in-the-bag @candy) 0)
+    (dosync
```

Comment:
Yeah, I know that it can't be summarised in two or three sentences but is a must at least to mention the STM. It's one of the core Clojure features, and then if you talk about state, immutability and concurrency I can't see how we cannot mention it. At least a link.